File: app/src/main/java/hr/caellian/notestream/gui/NoteStreamNotification.java
package hr.caellian.notestream.gui;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.BitmapFactory;
import android.os.Build;

import hr.caellian.notestream.gui.views.StatusBar;
import hr.caellian.notestream.gui.views.StatusBarLarge;
import hr.caellian.notestream.NoteStream;
import hr.caellian.notestream.R;
import hr.caellian.notestream.data.playable.Playable;
import hr.caellian.notestream.lib.Constants;
import hr.caellian.notestream.util.RepeatState;

/**
 * Created by caellyan on 24/06/17.
 */

public class NoteStreamNotification implements Playable.ControlListener {

    private Notification notification;

    StatusBar statusBar;
    StatusBarLarge statusBarLarge;

    Service parentService;
    Context context;

    boolean playing;

    NotificationManager nm;
    private final PendingIntent pendingIntent;
    Notification.Builder builder;

    public NoteStreamNotification(Context context, Service parentService) {
        String packageName = context.getPackageName();

        this.parentService = parentService;
        this.context = context;
        this.nm = (NotificationManager) context.getSystemService(Context.NOTIFICATION_SERVICE);

        statusBar = new StatusBar(context, packageName);
        statusBarLarge = new StatusBarLarge(context, packageName);

        Intent nsIntent = new Intent(context, ActivityLibrary.class);
        pendingIntent = PendingIntent.getActivity(context, Constants.APP_NOTIFICATION_ID, nsIntent, PendingIntent.FLAG_UPDATE_CURRENT);

        builder = new Notification.Builder(context);
        builder.setContentTitle(context.getString(R.string.app_name));

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            builder.setCustomContentView(statusBar);
            builder.setCustomBigContentView(statusBarLarge);
        } else {
            builder.setContent(statusBar);
        }

        builder.setContentIntent(pendingIntent);
        builder.setSmallIcon(R.drawable.ic_service_icon);
        builder.setLargeIcon(BitmapFactory.decodeResource(context.getResources(), R.drawable.ic_service_icon));

        NoteStream.registerControlListener(this);
    }

    public NoteStreamNotification rebuild() {
        notification = builder.build();
        if (playing) notification.flags |= Notification.FLAG_NO_CLEAR;
        notification.flags |= Notification.FLAG_ONGOING_EVENT;

        if (playing) {
            parentService.startForeground(Constants.APP_NOTIFICATION_ID, notification);
        } else {
            parentService.stopForeground(false);
            nm.notify(Constants.APP_NOTIFICATION_ID, notification);
        }
        return this;
    }

    public Notification getNotification() {
        return notification;
    }

    @Override
    public void onPlayableChanged(Playable current) {
        statusBar.onPlayableChanged(current);
        statusBarLarge.onPlayableChanged(current);
        rebuild();
    }

    @Override
    public void onPlayStatusChanged(boolean playing) {
        this.playing = playing;
        statusBar.onPlayStatusChanged(playing);
        statusBarLarge.onPlayStatusChanged(playing);
        rebuild();
    }

    @Override
    public void onShuffleStateChanged(boolean currentState) {
        statusBarLarge.onShuffleStateChanged(currentState);
        rebuild();
    }

    @Override
    public void onRepeatStateChanged(RepeatState currentState) {
        statusBarLarge.onRepeatStateChanged(currentState);
        rebuild();
    }
}

File: app/src/main/java/hr/caellian/notestream/gui/NoteStreamNotification.kt
/*
 * Copyright (C) 2018 Tin Svagelj
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
 * persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT
 * NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
 * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package hr.caellian.notestream.gui

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import hr.caellian.notestream.NoteStream
import hr.caellian.notestream.R
import hr.caellian.notestream.data.playable.Playable
import hr.caellian.notestream.gui.views.StatusBar
import hr.caellian.notestream.gui.views.StatusBarLarge
import hr.caellian.notestream.lib.Constants
import hr.caellian.notestream.util.RepeatState

/**
 * Created by caellyan on 24/06/17.
 */

class NoteStreamNotification(internal var context: Context, internal var parentService: Service) : Playable.ControlListener {

    var notification: Notification? = null
        private set

    internal var statusBar: StatusBar
    internal var statusBarLarge: StatusBarLarge

    internal var playing: Boolean = false

    internal var nm: NotificationManager
    private val pendingIntent: PendingIntent
    internal var builder: Notification.Builder

    init {
        val packageName = context.packageName
        this.nm = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        statusBar = StatusBar(context, packageName)
        statusBarLarge = StatusBarLarge(context, packageName)

        val nsIntent = Intent(context, ActivityLibrary::class.java)
        pendingIntent = PendingIntent.getActivity(context, Constants.APP_NOTIFICATION_ID, nsIntent, PendingIntent.FLAG_UPDATE_CURRENT)

        builder = Notification.Builder(context)
        builder.setContentTitle(context.getString(R.string.app_name))

        builder.setCustomContentView(statusBar)
        builder.setCustomBigContentView(statusBarLarge)

        builder.setContentIntent(pendingIntent)
        builder.setSmallIcon(R.drawable.ic_play_arrow)
        builder.setLargeIcon(BitmapFactory.decodeResource(context.resources, R.drawable.ic_play_arrow))

        NoteStream.registerControlListener(this)
    }

    fun rebuild(): NoteStreamNotification {
        notification = builder.build()
        parentService.startForeground(Constants.APP_NOTIFICATION_ID, notification)
        return this
    }

    fun cancelNotification() {
        NoteStream.CONTROL_LISTENERS.remove(this)
        parentService.stopForeground(true)
    }

    override fun onPlayableChanged(current: Playable?) {
        statusBar.onPlayableChanged(current)
        statusBarLarge.onPlayableChanged(current)
        rebuild()
    }

    override fun onPlayStatusChanged(playing: Boolean) {
        this.playing = playing
        statusBar.onPlayStatusChanged(playing)
        statusBarLarge.onPlayStatusChanged(playing)
        rebuild()
    }

    override fun onShuffleStateChanged(currentState: Boolean) {
        statusBarLarge.onShuffleStateChanged(currentState)
        rebuild()
    }

    override fun onRepeatStateChanged(currentState: RepeatState) {
        statusBarLarge.onRepeatStateChanged(currentState)
        rebuild()
    }
}
